lib/THC/THCReduce.cuh

```diff
@@ -11,24 +11,65 @@
 #include "THCTensorTypeUtils.cuh"
 #include "THCReduceApplyUtils.cuh"
 
-// Threads per thread block
-#define THC_NONCONTIG_REDUCE_BLOCK_SIZE 32 * 16
+#include <cstdio>
 
-template <typename IndexType>
-__device__ __forceinline__ IndexType getReduceNoncontigDimSliceIndex() {
-  // Each thread handles one slice
-  return getLinearBlockId<IndexType>() * THC_NONCONTIG_REDUCE_BLOCK_SIZE + threadIdx.x;
+#define LOCAL_MAX_BLOCK_SIZE 512
+
+template <typename ModifyOp,
+          typename ReduceOp,
+          typename T,
+          typename IndexType,
+          int ADims, int BDims>
+__global__ void
+kernelReduceNoncontigDim_shared(TensorInfo<T, IndexType> out,
+                         TensorInfo<T, IndexType> in,
+                         IndexType reductionStride,
+                         IndexType reductionSize,
+                         IndexType totalSlices,
+                         T init,
+                         ModifyOp modifyOp,
+                         ReduceOp reduceOp) {
+  IndexType threadLane  = threadIdx.x;
+  IndexType groupID      = threadIdx.y;
+  IndexType sliceIndex  = blockIdx.x * blockDim.x + threadLane;
+  IndexType sliceStride = gridDim.x * blockDim.x;
+
+  __shared__ T local_reduce[LOCAL_MAX_BLOCK_SIZE];
+  //If multiple groups of threads work on the same reduction use shared memory
+  //Otherwise use a register for reduction as only one thread is doing reduction
+  T* shmem = &local_reduce[threadIdx.x + threadIdx.y * blockDim.x];
+
+  for(;sliceIndex<totalSlices; sliceIndex+=sliceStride){
+    //  while(sliceIndex < totalSlices){
+    *shmem = init;
+
+    IndexType outOffset = IndexToOffset<T, IndexType, ADims>::get(sliceIndex, out);
+    IndexType inOffset = IndexToOffset<T, IndexType, BDims>::get(sliceIndex, in);
+    IndexType stride = reductionStride * blockDim.y;
+
+    for(IndexType i=groupID; i<reductionSize; i+=blockDim.y){
+      (*shmem) = reduceOp(*shmem, modifyOp(in.data[inOffset]) );
```

Comment:
I'm not sure exactly how this works. I could be wrong, but aren't we hitting shared memory every time here? If we want different threads in the "group" to reduce things in registers wouldn't we need a local variable?